File: zeroMQ/python/sender_test.py
#!/usr/bin/env python3
"""ZeroMQ Python Sender - Sync"""
import sys
import json
import zmq
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

from message_helpers import *
from test_data_loader import load_test_data
from stats_collector import MessageStats


def main():
    test_data = load_test_data()
    
    stats = MessageStats()
    stats.set_metadata({
        'service': 'ZeroMQ',
        'language': 'Python',
        'async': False
    })
    start_time = get_current_time_ms()
    
    print(f" [x] Starting transfer of {len(test_data)} messages...")
    
    context = zmq.Context()
    socket = context.socket(zmq.REQ)
    
    # Configure socket
    socket.setsockopt(zmq.RCVTIMEO, 100) # 100ms timeout
    socket.setsockopt(zmq.LINGER, 0)
    
    current_target = -1
    
    for item in test_data:
        message_id = extract_message_id(item)
        target = item.get('target', 0)
        
        # Connect to correct target port if changed
        if target != current_target:
             # Disconnect previous if needed? ZMQ handles multi-connect but REQ expects 1 reply per request
             # For simplicity, we can recreate socket or connect to all
             # But strictly speaking, REQ socket connecting to multiple endpoints does load balancing
             # To target specific receiver, we need specific connection
             
             socket.close()
             socket = context.socket(zmq.REQ)
             socket.setsockopt(zmq.RCVTIMEO, 100)
             socket.connect(f"tcp://localhost:{5556 + target}")
             current_target = target
             
        print(f" [x] Sending message {message_id} to target {target}...", end='', flush=True)
        msg_start = get_current_time_ms()
        
        # Create and send protobuf message
        envelope = create_data_envelope(item)
        body = serialize_envelope(envelope)
        
        socket.send(body)
        
        try:
            response = socket.recv()
            resp_envelope = parse_envelope(response)
            
            if is_valid_ack(resp_envelope, message_id):
                msg_duration = get_current_time_ms() - msg_start
                stats.record_message(True, msg_duration)
                print(" [OK]")
            else:
                stats.record_message(False)
                print(" [FAILED] Invalid ACK")
                
        except zmq.Again:
            stats.record_message(False)
            print(" [FAILED] Timeout")
            # Recreate socket on timeout to clear state
            socket.close()
            socket = context.socket(zmq.REQ)
            socket.setsockopt(zmq.RCVTIMEO, 100)
            socket.connect(f"tcp://localhost:{5556 + target}")
            
        except Exception as e:
            stats.record_message(False)
            print(f" [FAILED] {e}")
            
    socket.close()
    context.term()
    
    end_time = get_current_time_ms()
    stats.set_duration(start_time, end_time)
    
    report = stats.get_stats()
    
    print("\nTest Results:")
    print(f"service: ZeroMQ")
    print(f"language: Python")
    print(f"total_sent: {stats.sent_count}")
    print(f"total_received: {stats.received_count}")
    print(f"duration_ms: {stats.get_duration_ms()}")
    
    with open('logs/report.txt', 'a') as f:
        f.write(json.dumps(report) + '\n')


if __name__ == "__main__":
    main()

File: zeroMQ/cpp/sender_test.cpp
#include <zmq.hpp>
#include <string>
#include <iostream>
#include <fstream>
#include <chrono>
#include <thread>
#include <map>
#include "../../utils/cpp/stats_collector.hpp"
#include "../../utils/cpp/test_data_loader.hpp"
#include "../../utils/cpp/message_helpers.hpp"

using json = nlohmann::json;
using message_helpers::get_current_time_ms;

int main() {
    zmq::context_t context(1);
    std::map<int, zmq::socket_t*> sockets;

    auto test_data = test_data_loader::loadTestData();

    MessageStats stats;
    stats.set_metadata({
        {"service", "ZeroMQ"},
        {"language", "C++"},
        {"async", false}
    });
    long long start_time = get_current_time_ms();

    std::cout << " [x] Starting transfer of " << test_data.size() << " messages..." << std::endl;

    for (auto& item : test_data) {
        std::string message_id = message_helpers::extract_message_id(item);
        int target = item.value("target", 0);
        int port = 5556 + target;

        // Get or create socket for this target
        if (sockets.find(target) == sockets.end()) {
            zmq::socket_t* sock = new zmq::socket_t(context, ZMQ_REQ);
            sock->connect("tcp://localhost:" + std::to_string(port));
            sock->setsockopt(ZMQ_RCVTIMEO, 40);  // 40ms timeout
            sockets[target] = sock;
            
            // Small delay to allow ZeroMQ connection to establish
            // (connect() is async and returns immediately)
            // Async receivers need slightly more time
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }
        
        zmq::socket_t* socket = sockets[target];
        std::cout << " [x] Sending message " << message_id << " to port " << port << "..." << std::flush;
        
        long long msg_start = get_current_time_ms();
        try {
            // Create and send protobuf message
            MessageEnvelope envelope = message_helpers::create_data_envelope(item);
            std::string body = message_helpers::serialize_envelope(envelope);
            
            zmq::message_t request(body.size());
            memcpy(request.data(), body.c_str(), body.size());
            socket->send(request, zmq::send_flags::none);

            // Receive ACK
            zmq::message_t reply;
            auto recv_res = socket->recv(reply);
            
            if (recv_res.has_value()) {
                std::string reply_str(static_cast<char*>(reply.data()), reply.size());
                
                // Parse and validate ACK
                MessageEnvelope resp_envelope;
                if (message_helpers::parse_envelope(reply_str, resp_envelope) && 
                    message_helpers::is_valid_ack(resp_envelope, message_id)) {
                    long long msg_duration = get_current_time_ms() - msg_start;
                    stats.record_message(true, msg_duration);
                    std::cout << " [OK]" << std::endl;
                } else {
                    stats.record_message(false);
                    std::cout << " [FAILED] Invalid ACK" << std::endl;
                }
            } else {
                stats.record_message(false);
                std::cout << " [FAILED] Timeout" << std::endl;
                // Close poisoned REQ socket
                socket->close();
                delete socket;
                sockets.erase(target);
            }

        } catch (const std::exception& e) {
            stats.record_message(false);
            std::cout << " [FAILED] Error: " << e.what() << std::endl;
            socket->close();
            delete socket;
            sockets.erase(target);
        }
    }

    long long end_time = get_current_time_ms();
    stats.set_duration(start_time, end_time);
    
    json report = stats.get_stats();

    std::cout << "\nTest Results:" << std::endl;
    std::cout << "service: ZeroMQ" << std::endl;
    std::cout << "language: C++" << std::endl;
    std::cout << "total_sent: " << stats.sent_count << std::endl;
    std::cout << "total_received: " << stats.received_count << std::endl;
    std::cout << "duration_ms: " << stats.get_duration_ms() << std::endl;

    std::ofstream rf("logs/report.txt", std::ios::app);
    if (rf.good()) {
        rf << report.dump() << std::endl;
        rf.close();
    }

    // Cleanup
    for (auto& pair : sockets) {
        pair.second->close();
        delete pair.second;
    }

    return 0;
}
